docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
```

Comment:
Define meta-discovery API?  Adding it to /apis/ when registered would
satisfy I think the bulk of the clients.

ThirdPartyResource would have to be given their own APIGroupLocation
probably.

One component may represent multiple apigrouplocations.

On Tue, Nov 17, 2015 at 5:31 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/third-party-api-servers.md
> <https://github.com/kubernetes/kubernetes/pull/17191#discussion_r45132497>
> :
>
> > +expose new APIs at runtime without having to modify the core kubernetes API
> > +server binary.
> > +Anyone can write their own API server and register it with the core
> > +kubernetes API server. These new APIs will be seamless extension of core
> > +kubernetes APIs (ex: they will be operated upon via kubectl).
> > +
> > +## Non Goals:
> > +
> > +The following are related but are not the goals of this specific proposal:
> > +* Make it easy to write a kubernetes API server.
> > +
> > +## High Level Architecture
> > +
> > +Users can bring up their own API servers and register them with the core server
> > +at runtime.
> > +Core server will then start exposing these third party servers in the discovery
>
> Do we want/need this, or would a meta-discovery API be better?
>
> Also, how does this relate to:
>
>    - the component registration API
>    - ThirdPartyResource
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17191/files#r45132497>.
>